### PR TITLE
feat(config): adds override_empty_host_header flag

### PR DESCRIPTION
**Pull request by M4tteoP**
The feature added in https://github.com/coreruleset/go-ftw/pull/63 replaces the empty host field with the destination address. It has conflicts with some CRS test that are trying to send on purpose empty or missing host headers.
It happens only when `dest_addr` override is in place, for this reason CRS tests against apache are still running smoothly.
This PR implements @theseion [idea](https://github.com/coreruleset/go-ftw/issues/129#issuecomment-1421347248), adding the `override_empty_host_header` flag that enables this feature, permitting, by default, to keep running CRS tests as intended. 
More context in the issue: https://github.com/coreruleset/go-ftw/issues/129.

Thanks the feedbacks!

tentatively closes https://github.com/coreruleset/go-ftw/issues/129

### CODE REVIEW

**Review by jcchavezs — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -297,6 +297,7 @@ Sometimes you have tests that work well for some platform combinations, e.g. Apa
 
 Tests can be altered using four lists:
 - `input` allows you to override global parameters in tests. An example usage is if you want to change the `dest_addr` of all tests to point to an external IP or host
+  - it includes also the `override_empty_host_header` flag: enabling it, empty Host headers will be replaced with `dest_addr`
```
Comment (jcchavezs) [inline README.md:300]: It feels like a breaking change. Are we OK with that? Shall we do it the other way around? Just loud thoughts.

Comment (M4tteoP) [reply README.md:300]: Being an override, part of the optional `testoverride` options, I felt that the correct default option would be to not override. Writing a test with an empty Host, I'm expecting that the Host is indeed sent empty.
If it sounds bad, I can turn it into `not_override_empty_host_header`

Comment (theseion) [inline README.md:300]: I think we should explicitly document all available overrides. That will make this section clearer.

Comment (theseion) [reply README.md:300]: Yes, it would be breaking, but I agree with @M4tteoP. I only know of one person who uses this feature and I don't think it would be too much to ask to make a small adjustment. We'll increase the major version number, that should suffice as a signal. @fzipi what do you think?

File: runner/run_test.go [hunk 2]
```diff
@@ -675,7 +675,7 @@ func TestFailedTestsRun(t *testing.T) {
 	assert.Equal(t, 1, res.Stats.TotalFailed())
 }
 
-func TestApplyInputOverrideSetHostFromDestAddr(t *testing.T) {
+func TestApplyInputOverrideEmptyHostHeaderSetHostFromDestAddr(t *testing.T) {
```
Comment (theseion) [inline runner/run_test.go:678]: Add another test for the default case, where `OverrideEmptyHostHeader` has not been set.

File: test/types.go [hunk 3]
```diff
@@ -5,18 +5,19 @@ import "github.com/coreruleset/go-ftw/ftwhttp"
...
+	URI                     *string        `yaml:"uri,omitempty" koanf:"uri,omitempty"`
+	Version                 *string        `yaml:"version,omitempty" koanf:"version,omitempty"`
+	Headers                 ftwhttp.Header `yaml:"headers,omitempty" koanf:"headers,omitempty"`
+	Method                  *string        `yaml:"method,omitempty" koanf:"method,omitempty"`
+	Data                    *string        `yaml:"data,omitempty" koanf:"data,omitempty"`
+	SaveCookie              bool           `yaml:"save_cookie,omitempty" koanf:"save_cookie,omitempty"`
+	StopMagic               bool           `yaml:"stop_magic" koanf:"stop_magic,omitempty"`
+	EncodedRequest          string         `yaml:"encoded_request,omitempty" koanf:"encoded_request,omitempty"`
+	RAWRequest              string         `yaml:"raw_request,omitempty" koanf:"raw_request,omitempty"`
+	OverrideEmptyHostHeader bool           `yaml:"override_empty_host_header,omitempty" koanf:"override_empty_host_header,omitempty"`
```
Comment (theseion) [inline test/types.go:20]: I wonder whether we should start using a separate struct for the overrides. The new option has no place in the `input` map of tests.

**Review by theseion — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (M4tteoP) [reply README.md]: Done, I tried to write a small description of each, but not sure about `savecookie` and `stopmagic` 🤔

File: README.md [hunk 4]
```diff
@@ -296,7 +296,20 @@ Other interesting functions you can use are: `randBytes`, `htpasswd`, `encryptAE
 Sometimes you have tests that work well for some platform combinations, e.g. Apache + modsecurity2, but fail for others, e.g. NGiNX + modsecurity3. Taking that into account, you can override test results using the `testoverride` config param. The test will be skipped, and the result forced as configured.
 
 Tests can be altered using four lists:
-- `input` allows you to override global parameters in tests. An example usage is if you want to change the `dest_addr` of all tests to point to an external IP or host
+- `input` allows you to override global parameters in tests. The following ones can be overridden:
+  - `dest_addr`: allows to point to an external IP or host
+  - `override_empty_host_header`: enabling this flag, empty Host headers will be replaced with `dest_addr`
```
Comment (theseion) [inline README.md:301]:
```suggestion
  - `override_empty_host_header`: if true and `dest_addr` override is _also_ set, empty `Host` headers will be replaced with `dest_addr`
```

File: README.md [hunk 5]
```diff
@@ -296,7 +296,20 @@ Other interesting functions you can use are: `randBytes`, `htpasswd`, `encryptAE
 Sometimes you have tests that work well for some platform combinations, e.g. Apache + modsecurity2, but fail for others, e.g. NGiNX + modsecurity3. Taking that into account, you can override test results using the `testoverride` config param. The test will be skipped, and the result forced as configured.
 
 Tests can be altered using four lists:
-- `input` allows you to override global parameters in tests. An example usage is if you want to change the `dest_addr` of all tests to point to an external IP or host
+- `input` allows you to override global parameters in tests. The following ones can be overridden:
+  - `dest_addr`: allows to point to an external IP or host
```
Comment (theseion) [inline README.md:300]:
```suggestion
  - `dest_addr`: overrides the destination address (accepts IP or hostname)
```

File: README.md [hunk 6]
```diff
@@ -296,7 +296,20 @@ Other interesting functions you can use are: `randBytes`, `htpasswd`, `encryptAE
...
+  - `override_empty_host_header`: enabling this flag, empty Host headers will be replaced with `dest_addr`
+  - `port`: overrides the port number
+  - `protocol`: overrides the protocol
+  - `uri`: overrides the uri
+  - `version`: overrides the HTTP version. E.g. "HTTP/1.1"
+  - `headers`: overrides headers, the format is a map of strings
+  - `method`: overrides the method used to perform the request
+  - `data`: overrides data sent in the request
+  - `savecookie`: 
+  - `stopmagic`:
```
Comment (theseion) [inline README.md:310]: If we keep this option, we should rename it IMO (keep it for backward compatibility but add a sibling that has a better name and document that here). @fzipi?

```suggestion
  - `stopmagic`: prevent header autocompletion (currently sets `Connection: close` and `Content-Length` for requests with body data)
```

File: README.md [hunk 7]
```diff
@@ -296,7 +296,20 @@ Other interesting functions you can use are: `randBytes`, `htpasswd`, `encryptAE
...
+  - `dest_addr`: allows to point to an external IP or host
+  - `override_empty_host_header`: enabling this flag, empty Host headers will be replaced with `dest_addr`
+  - `port`: overrides the port number
+  - `protocol`: overrides the protocol
+  - `uri`: overrides the uri
+  - `version`: overrides the HTTP version. E.g. "HTTP/1.1"
+  - `headers`: overrides headers, the format is a map of strings
+  - `method`: overrides the method used to perform the request
+  - `data`: overrides data sent in the request
+  - `savecookie`:
```
Comment (theseion) [inline README.md:309]: Not yet implemented. @fzipi any plans to do that?

**Review by theseion — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (theseion) [reply README.md]: Thanks a lot!

File: README.md (hunk 7 above)
Comment (M4tteoP) [reply README.md]: Many thanks for the suggested changes Max! Being not implemented I temptatively removed it from the list

Comment (fzipi) [reply README.md]: I think we need to uncomment https://github.com/coreruleset/go-ftw/blob/94ab311941820255097b13fe53c6f7276e19613a/ftwhttp/request.go#L165-L170 and provide some tests.

Comment (theseion) [reply README.md]: Ah yes. In that case @M4tteoP, I propose to drop `savecooke` for now, as you did. We can add it to the documentation once cookie handling has been implemented.

File: README.md (hunk 1 above)
Comment (fzipi) [reply README.md]: This is part of the test yaml, right? Then we cannot change it without a new version of the yaml spec, sorry.

Comment (M4tteoP) [reply README.md]: So should I switch it to `not_override_empty_host_header `?
I think that #63  broke the default behavior, and enabling the custom behavior of overriding empty headers via `override_empty_host_header` would restore it.
We would have `dest_addr` that just overrides the destination address and the `override_empty_host_header` flag that force to override also the empty host headers, without any implicit behavior.

Comment (theseion) [reply README.md]: I don't understand your answer @fzipi 😅. I vote for `override_empty_host_header` (i.e., the "breaking change").

File: README.md (hunk 6 above)
Comment (fzipi) [reply README.md]: Yes `stopmagic` is clearly not a good name regarding its semantic. We can use a better one.

Comment (theseion) [reply README.md]: I have a draft PR waiting to rename stopmagic after this PR has been merged.

File: README.md (hunk 1 above)
Comment (fzipi) [reply README.md]: I don't mind the breaking change, as this is only in go-ftw. We should bail with an error for users that had the previous behavior, right?

Comment (theseion) [reply README.md]: How would we do that? Wouldn't that still be a valid configuration (valid in the context of testing, not w.r.t. HTTP)?
- dest_addr override would be set
- Host header would be empty
- Host header override would be empty

If you wanted to test request spoofing, i.e., use a host that differs from the target address, then you can't raise an exception here, right?

File: test/types.go (hunk 3 above)
Comment (fzipi) [reply test/types.go:20]: I also think we should separate the overrides, yes.

**Review by fzipi — COMMENTED** [review]
If we move the overrides to a new struct it will simplify things in the future...

File: runner/run.go [hunk 8]
```diff
@@ -379,16 +379,26 @@ func getRequestFromTest(testRequest test.Input) *ftwhttp.Request {
 
 // applyInputOverride will check if config had global overrides and write that into the test.
 func applyInputOverride(o config.FTWTestOverride, testRequest *test.Input) error {
-	overrides := o.Input
+	overrides := o.Overrides
 	if overrides.Port != nil {
 		testRequest.Port = overrides.Port
 	}
+
+	if overrides.Headers != nil {
```
Comment (M4tteoP) [inline runner/run.go:387]: I added headers override support starting from @snake66's [PR](https://github.com/coreruleset/go-ftw/pull/144) (Thank you!). Specifically speaking about the Host header, the individual header override will take precedence over a host header specified in the test and over the OverrideEmptyHostHeader feature. I added a test about it.

File: test/types.go [hunk 9]
```diff
@@ -19,6 +20,23 @@ type Input struct {
 	RAWRequest     string         `yaml:"raw_request,omitempty" koanf:"raw_request,omitempty"`
 }
 
+// Overrides represents the overridden inputs that have to be applied to tests
+type Overrides struct {
```
Comment (M4tteoP) [inline test/types.go:24]: Little attempt to move the overrides to a new struct. When it will be settled, I will open another PR for extending the override capabilities (all the ones that I have commented out are currently not enforced)

File: runner/run_test.go [hunk 10]
```diff
@@ -742,11 +770,101 @@ func TestApplyInputOverrideSetHostFromDestAddr(t *testing.T) {
...
+
+	err := applyInputOverride(cfg.TestOverride, &testInput)
+	assert.NoError(t, err, "Failed to apply input overrides")
+
+	assert.Equal(t, overrideHost, *testInput.DestAddr, "`dest_addr` should have been overridden")
+
+	assert.NotNil(t, testInput.Headers, "Header map must exist after overriding `dest_addr`")
+
 	hostHeader := testInput.Headers.Get("Host")
 	assert.NotEqual(t, "", hostHeader, "Host header must be set after overriding `dest_addr`")
```
Comment (theseion) [inline runner/run_test.go:800]:
```suggestion
	assert.NotEqual(t, "", hostHeader, "Host header must be set after overriding `dest_addr` and setting `override_empty_host_header` to `true`")
```

File: runner/run_test.go [hunk 11]
```diff
@@ -742,11 +770,101 @@ func TestApplyInputOverrideSetHostFromDestAddr(t *testing.T) {
...
 	assert.Equal(t, overrideHost, hostHeader, "Host header must be identical to `dest_addr` after overrding `dest_addr`")
```
Comment (theseion) [inline runner/run_test.go:801]:
```suggestion
	assert.Equal(t, overrideHost, hostHeader, "Host header must be identical to `dest_addr` after overriding `dest_addr` and setting `override_emtpy_host_header` to `true`")
```

**Review by theseion — CHANGES_REQUESTED** [review]

**Review by theseion — APPROVED** [review]

**Comment by theseion** [comment]
Thanks @M4tteoP, looks good. I've requested a review from @fzipi, just to be sure we're all on the same page.

**Review by fzipi — APPROVED** [review]
Thanks for putting this together! LGTM.

**Comment by camelmasa** [comment]
Great work! If possible, Please release a new version 👍

**Comment by theseion** [comment]
https://github.com/coreruleset/go-ftw/releases/tag/v0.4.9

**Comment by camelmasa** [comment]
Thank you so much!
